google/cloud/internal/async_polling_loop_test.cc

```diff
@@ -234,31 +247,40 @@ TEST(AsyncPollingLoopTest, PollThenEventualSuccess) {
   EXPECT_CALL(*mock, AsyncGetOperation)
       .WillOnce([&](CompletionQueue&, std::unique_ptr<grpc::ClientContext>,
                     google::longrunning::GetOperationRequest const&) {
+        EXPECT_EQ(CurrentOptions().get<StringOption>(),
+                  "PollThenEventualSuccess");
         return make_ready_future(make_status_or(starting_op));
       })
       .WillOnce([](CompletionQueue&, std::unique_ptr<grpc::ClientContext>,
                    google::longrunning::GetOperationRequest const&) {
+        EXPECT_EQ(CurrentOptions().get<StringOption>(),
+                  "PollThenEventualSuccess");
         return make_ready_future(
             StatusOr<Operation>(Status(StatusCode::kUnavailable, "try-again")));
       })
       .WillOnce([&](CompletionQueue&, std::unique_ptr<grpc::ClientContext>,
                     google::longrunning::GetOperationRequest const&) {
+        EXPECT_EQ(CurrentOptions().get<StringOption>(),
+                  "PollThenEventualSuccess");
         return make_ready_future(make_status_or(starting_op));
       })
       .WillOnce([&](CompletionQueue&, std::unique_ptr<grpc::ClientContext>,
                     google::longrunning::GetOperationRequest const&) {
+        EXPECT_EQ(CurrentOptions().get<StringOption>(),
+                  "PollThenEventualSuccess");
         return make_ready_future(make_status_or(expected));
       });
   auto policy = absl::make_unique<MockPollingPolicy>();
   EXPECT_CALL(*policy, clone()).Times(0);
   EXPECT_CALL(*policy, OnFailure).WillRepeatedly(Return(true));
   EXPECT_CALL(*policy, WaitPeriod)
       .WillRepeatedly(Return(std::chrono::milliseconds(1)));
-  auto actual =
-      AsyncPollingLoop(cq, make_ready_future(make_status_or(starting_op)),
-                       MakePoll(mock), MakeCancel(mock), std::move(policy),
-                       "test-function")
-          .get();
+  OptionsSpan span(Options{}.set<StringOption>("PollThenEventualSuccess"));
+  auto pending = AsyncPollingLoop(
+      cq, make_ready_future(make_status_or(starting_op)), MakePoll(mock),
+      MakeCancel(mock), std::move(policy), "test-function");
+  OptionsSpan overlay(Options{}.set<StringOption>("uh-oh"));
```

Comment:
Unfortunately I think this test does not catch the problems because all the asynchronous callbacks run immediately.  I think you want to use an `AsyncSequencer<T>` to gain finer grained control of the steps in this test, and reset the Options span each time.  This outline will probably need some work:

```cc
TEST(AsyncPollingLoopTest, PollThenEventualSuccess) {
  // NEW
  AsyncSequencer<Status> sequencer;

  Instance instance;
  instance.set_name("test-instance-name");
  google::longrunning::Operation starting_op;
  starting_op.set_name("test-op-name");
  google::longrunning::Operation expected = starting_op;
  expected.set_done(true);
  expected.mutable_metadata()->PackFrom(instance);
  auto mock_cq = std::make_shared<MockCompletionQueueImpl>();
  EXPECT_CALL(*mock_cq, MakeRelativeTimer)
      .WillRepeatedly([&](std::chrono::nanoseconds) {
        return sequencer.PushBack("MakeRelativeTimer").then([](future<Status> f) {
          auto status = f.get();
          if (status.ok()) return make_status_or(std::chrono::system_clock::now());
          return status; // Maybe needs an explicit conversion?
      });
  CompletionQueue cq(mock_cq);

  auto mock = std::make_shared<MockStub>();
  EXPECT_CALL(*mock, AsyncGetOperation)
      .WillOnce([&](CompletionQueue&, std::unique_ptr<grpc::ClientContext>,
                    google::longrunning::GetOperationRequest const&) {
        EXPECT_EQ(CurrentOptions().get<StringOption>(),
                  "PollThenEventualSuccess");
        return sequencer.PushBack("GetOperation").then([&](future<Status> f) {
          auto status = f.get();
          if (status.ok()) return make_status_or(starting_op);
          return status; // Maybe needs an explicit conversion?
        });
      })
... ... ...

  auto policy = absl::make_unique<MockPollingPolicy>();
  EXPECT_CALL(*policy, clone()).Times(0);
  EXPECT_CALL(*policy, OnFailure).WillRepeatedly(Return(true));
  EXPECT_CALL(*policy, WaitPeriod)
      .WillRepeatedly(Return(std::chrono::milliseconds(1)));
  OptionsSpan span(Options{}.set<StringOption>("PollThenEventualSuccess"));
  auto pending = AsyncPollingLoop(
      cq, make_ready_future(make_status_or(starting_op)), MakePoll(mock),
      MakeCancel(mock), std::move(policy), "test-function");

  // Now you can control the execution of the polling loop one callback at a time.

  // TODO this may not be a good example, but a good start maybe, with some C++17 liberties
  auto [p, name] = sequencer.PopFrontWithName();
  EXPECT_EQ(name, "GetOperation");
  OptionsSpan overlay(Options{}.set<StringOption>("uh-oh"));
  p.set_value(Status(StatusCode::kUnavailable, "try-again"));

  [p, name] = sequencer.PopFrontWithName();
  EXPECT_EQ(CurrentOptions().get<StringOption>(), "uh-oh"); // TODO - remove this, just pointing out what I expect
  EXPECT_EQ(name, "MakeRelativeTimer");
  p.set_value(Status()); // timers should succeed or the loop is aborted 

    
}

```